### ISSUE TITLE
Add support for CNN and attention-lstm?

### ISSUE THREAD

**Issue by xinsuinizhuan** [josephjaspers/blackcat_tensors#28]
attention-lstm is better than lstm， and cnn is need for forecast to group cnn and（attention-lstm）net.

**Comment by xinsuinizhuan** [comment]
![net](https://user-images.githubusercontent.com/40679769/63312093-ea685d00-c332-11e9-8b7b-cd13448b7d4b.jpg)
attention-lstm and your old project's net

**Comment by josephjaspers** [comment]
I am not sure what an attention-lstm is could you send a link?
I will start working on the CNN.  

I need to implement/find a good implementation of convolution to use. (iI have written it myself before but it was very slow).

**Comment by xinsuinizhuan** [comment]
about lstm-attention:
https://www.sciencedirect.com/science/article/abs/pii/S0950705119302400
https://www.sciencedirect.com/science/article/abs/pii/S1359431118343485
https://aclweb.org/anthology/D16-1058
https://arxiv.org/pdf/1908.02252.pdf
https://machinelearningmastery.com/how-does-attention-work-in-encoder-decoder-recurrent-neural-networks/
https://www.depends-on-the-definition.com/attention-lstm-relation-classification/
https://github.com/ningshixian/LSTM_Attention

**Comment by xinsuinizhuan** [comment]
the Long Short Term Memory Fully Convolutional Network (LSTM-FCN) and Attention LSTM-FCN (ALSTM-FCN)：

https://www.sciencedirect.com/science/article/pii/S0893608019301200?via%3Dihub
https://www.sciencedirect.com/science/article/pii/S0378437118314985?dgcid=raven_sd_recommender_email
https://www.sciencedirect.com/user/recommendations

**Comment by josephjaspers** [comment]
Added CNN as of:

https://github.com/josephjaspers/blackcat_tensors/commit/bbc1a2a49452b87f60fba15b2fc28958b4e0034d

It is very slow, and the user must calculate the output-shape itself. 
It is "experimental" currently but I will be completing within the next week. 
(Improving performance, auto calculating shape, etc).

**Comment by josephjaspers** [comment]
TODO:

Improve implementation of CNN (most likely switch to img2col implementation). 
Add Maxpooling
Add GPU Support. 
Add auto-deduction of the shape. 
Add Attention-LSTM

**Comment by xinsuinizhuan** [comment]
you are so great.I expect it so much.

**Comment by xinsuinizhuan** [comment]
How about Maxpooling? when use the cnn, shoud add Maxpooling?

**Comment by josephjaspers** [comment]
I have not yet implemented max-pooling, I want to try to optimize CNN first, but maxpooling isn't particularly difficult to implement so I can see if I can do that quickly.

**Comment by xinsuinizhuan** [comment]
> 
> 
> How about Maxpooling,when use the cnn, shoud add Maxpooling? When I want try to use this truct model, how should i do?
> ![图片](https://user-images.githubusercontent.com/40679769/65847174-03711e80-e373-11e9-8ec6-54e7c3e04411.png)

How about this net struct, now can implement?

**Comment by josephjaspers** [comment]
Yes I can work on that soon. 
I am also working on optimizing LSTM and the Convolution Layer. 

For Convolution and Maxpooling I may borrow Caffe's implementation.

**Comment by josephjaspers** [comment]
You can see that I have started working on max-pooling here: https://github.com/josephjaspers/blackcat_tensors/blob/master/include/neural_networks/functions/Max_Pooling.h

**Comment by xinsuinizhuan** [comment]
Yes. You are so great. I have seen the Max_Pooling.h, so i ask you that. Now we should first focus on the lstm single_predict, the same input and output, then to implement Convolution and Maxpooling.I expect it.

**Comment by xinsuinizhuan** [comment]
Because i seen one paper, that net struct is efficient to make my forcast. Now only lstm layer the forcast result is not so good, we should make other net sturct or use the attention-lstm.But that is next to implement and test.

**Comment by josephjaspers** [comment]
The order of things I will work on is...

1. Optimizing Convolution (It is too slow)
2. MaxPooling
3. AttentionLSTM 

I have found Caffe's implementation of Convoltution/MaxPooling so I will most likely be importing their implementation into this project.

**Comment by xinsuinizhuan** [comment]
thank you very much! But  i  hope  first  to debug  the single_predict function,how  about  it  now? four days  have no update.

在 2019-11-02 06:14:36，"Joseph Jaspers" <notifications@github.com> 写道：


The order of things I will work on is...

Optimizing Convolution (It is too slow)
MaxPooling
AttentionLSTM

I have found Caffe's implementation of Convoltution/MaxPooling so I will most likely be importing their implementation into this project.

**Comment by josephjaspers** [comment]
I am working on improving Convolution, hopefully I will have it finished by today or tomorrow.
I will try fix single_predict soon after.

**Comment by josephjaspers** [comment]
Here I just added a new version of Convolution. (It still needs testing, does not support single-predict currently).

https://github.com/josephjaspers/blackcat_tensors/blob/master/include/neural_networks/Layers/Convolution_Experimental.h

However it should be much faster than the current version. 
I will look into the single_predict function now.

Then...
--> Testing/adding single_predict to the faster Convolution method. 
--> Max_Pooling 
--> Optimizing LSTM
--> Adding Attention-LSTM

**Comment by josephjaspers** [comment]
> Because i seen one paper, that net struct is efficient to make my forcast. Now only lstm layer the forcast result is not so good, we should make other net sturct or use the attention-lstm.But that is next to implement and test.

I also just fixed a bug where set_learning_rate wouldn't actually set the learning_rate of the Layer. So perhaps re-running may improve performance.

**Comment by xinsuinizhuan** [comment]
You are so great. I am so glad to see the updata. I will test the new code just right.So begin our work as you plan.

**Comment by xinsuinizhuan** [comment]
> 
> 
> > Because i seen one paper, that net struct is efficient to make my forcast. Now only lstm layer the forcast result is not so good, we should make other net sturct or use the attention-lstm.But that is next to implement and test.
> 
> I also just fixed a bug where set_learning_rate wouldn't actually set the learning_rate of the Layer. So perhaps re-running may improve performance.

Yes, it is better in performance.But it need train more epochs:
befor version:
BC::nn::RMSE loss, begin is 0.156, i set epoch == 1028, then it reduce to 0.12, and have many mutations in the data.
now version:
BC::nn::RMSE loss, begin is 0.256, i set epoch == 5000, then it reduce to 0.07, and the data is relatively stable. When i set epoch == 1024, it only reduce to 0.166.

**Comment by josephjaspers** [comment]
Eventually I would like to add optimizers (like momentum and Adam), though I havent begun to work on them yet.

**Comment by xinsuinizhuan** [comment]
> 
> 
> Eventually I would like to add optimizers (like momentum and Adam), though I havent begun to work on them yet.

I am glad to see your reply! I am so expected. I  want to use the net in practical, so you should stepping up time, when you no busy!

**Comment by josephjaspers** [comment]
Convolution (the experimental version) is now the standard version. 
I have tested it (not windows but on linux). 
It is considerably faster than the previous version, however it consumes a lot of memory. 

https://github.com/josephjaspers/blackcat_tensors/commit/ff33dff0c4b5ec1f2231ccc941d42817e9b87415

**Comment by josephjaspers** [comment]
Maxpooling branch. (Not complete)

https://github.com/josephjaspers/blackcat_tensors/commit/5e6602fcdf363f1f2b007fb7de61e28cf83d383d

**Comment by josephjaspers** [comment]
convolution and maxpooling, have been added!
The attention-lstm ticket has been moved to: 
https://github.com/josephjaspers/blackcat_tensors/issues/48
